src/app/clusters/access-control-server/access-control-server.cpp

```diff
@@ -426,68 +426,84 @@ CHIP_ERROR AccessControlAttribute::Write(const ConcreteDataAttributePath & aPath
-    ReturnErrorOnFailure(GetAccessControl().GetMaxEntryCount(maxCount));
-    VerifyOrReturnError(allCount >= oldCount, CHIP_ERROR_INTERNAL);
-    VerifyOrReturnError(static_cast<size_t>(allCount - oldCount + newCount) <= maxCount, CHIP_ERROR_INVALID_LIST_LENGTH);
+        ReturnErrorOnFailure(GetAccessControl().GetEntryCount(allCount));
+        ReturnErrorOnFailure(list.ComputeSize(&newCount));
+        ReturnErrorOnFailure(GetAccessControl().GetMaxEntryCount(maxCount));
+        VerifyOrReturnError(allCount >= oldCount, CHIP_ERROR_INTERNAL);
+        VerifyOrReturnError(static_cast<size_t>(allCount - oldCount + newCount) <= maxCount, CHIP_ERROR_INVALID_LIST_LENGTH);
 
-    auto iterator = list.begin();
-    size_t i      = 0;
-    while (iterator.Next())
-    {
-        if (i < oldCount)
+        auto iterator = list.begin();
+        size_t i      = 0;
+        while (iterator.Next())
         {
-            ReturnErrorOnFailure(GetAccessControl().UpdateEntry(i, iterator.GetValue().entry, &accessingFabricIndex));
-            ReturnErrorOnFailure(LogAccessControlEvent(iterator.GetValue().entry, aDecoder.GetSubjectDescriptor(),
-                                                       AccessControlCluster::ChangeTypeEnum::kChanged));
+            if (i < oldCount)
+            {
+                ReturnErrorOnFailure(GetAccessControl().UpdateEntry(i, iterator.GetValue().entry, &accessingFabricIndex));
+                ReturnErrorOnFailure(LogAccessControlEvent(iterator.GetValue().entry, aDecoder.GetSubjectDescriptor(),
+                                                           AccessControlCluster::ChangeTypeEnum::kChanged));
+            }
+            else
+            {
+                ReturnErrorOnFailure(GetAccessControl().CreateEntry(nullptr, iterator.GetValue().entry, &accessingFabricIndex));
+                ReturnErrorOnFailure(LogAccessControlEvent(iterator.GetValue().entry, aDecoder.GetSubjectDescriptor(),
+                                                           AccessControlCluster::ChangeTypeEnum::kAdded));
+            }
+            ++i;
         }
-        else
+        ReturnErrorOnFailure(iterator.GetStatus());
+
+        while (i < oldCount)
         {
-            ReturnErrorOnFailure(GetAccessControl().CreateEntry(nullptr, iterator.GetValue().entry, &accessingFabricIndex));
+            --oldCount;
+            ReturnErrorOnFailure(GetAccessControl().DeleteEntry(oldCount, &accessingFabricIndex));
             ReturnErrorOnFailure(LogAccessControlEvent(iterator.GetValue().entry, aDecoder.GetSubjectDescriptor(),
-                                                       AccessControlCluster::ChangeTypeEnum::kAdded));
+                                                       AccessControlCluster::ChangeTypeEnum::kRemoved));
         }
-        ++i;
     }
-    ReturnErrorOnFailure(iterator.GetStatus());
+    else if (aPath.mListOp == ConcreteDataAttributePath::ListOperation::AppendItem)
```

Comment:
This new clause seems correct to me: decode item into codec, create it on the accessing fabric, log it.